20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,489 @@
+
+When the time comes, check will be updated to ignore tuples that do not match the assignable tuples in the authorization model.
+
+## Migration
+
+### Migrating existing authorization models
+
+Existing models will not be migrated - new models will be required to use types (with an optional grace period). Authorization models that do not make use of types will not be able to use the optimized ListObjects endpoint.
+
+### Migrating existing tuples
+
+Due to the users in the tuples now required to have types in order to enforce the restrictions, existing tuples with users as floating user identifiers with no types will no longer be valid when types are added. These tuples will not be removed from the system, and will still be valid while the legacy authorization model is supported, but will be ignored during evaluation on newer authorization models with type restrictions in place.
+
+This will need to be communicated to users so that they can migrate accordingly.
+
+One option could be that a script to check the DB (look for tuples in the DB that do not have `user_type`), and prints the offending store IDs.
+
+Automatic unassisted tuple migration will not be feasible.
+
+### Affected Modules
+
+This change will affect repositories across the board. It will entail changes to the [protobuf](https://github.com/openfga/api), the [openfga core](https://github.com/openfga/openfga), the DSL, the [syntax transformer](https://github.com/openfga/syntax-transformer), the [SDKs](https://github.com/openfga/sdk-generator), the FGA Playground, the [sample stores](https://github.com/openfga/sample-stores) and the [documentation](https://github.com/openfga/openfga.dev).
+
+#### Language, API and SDKs
+
+For the scope of this RFC, we are proposing that the initial phase be backwards compatible and not a breaking change. This will allow users on previous versions to keep using them for a while, similarly changes to the public surface of the API and SDKs are reduced.
+
+A later RFC can introduce the upgraded syntax where the extension now placed in metadata can be migrated. That phase can be combined with other breaking changes we are introducing to minimize user-disruption.
+
+#### DSL
+
+An update to teh DSL needs to be drafted to support the inclusion of the type restrictions, this should come in a later RFC.
+
+#### Syntax Transformer and Playground
+
+Syntax transformer will need to be updated to support both the new JSON syntax and the new DSL it maps to, as well as all the necessary validations.
+
+### Roadmap
+
+1. This RFC is drafted
+2. The [openfga/api#27](https://github.com/openfga/api/pull/27) PR introduces the new fields into the proto-files
+3. The DSL & [openfga/syntax-transformer](https://github.com/openfga/syntax-transformer) are updated to allow empty user type
```

Comment:
This was not part of this RFC 🤔

Reply:
Added a note on this in the DSL migration above